### PR TITLE
First Major PR

### PR DESCRIPTION
**Pull request by yinghang**
## Overview
This module aims to expose the ability to determine whether the audio jack is plugged in in React Native land for both iOS & Android.

## Changes
Methods provided via this module are as follows:

```
RNAudioJack {
  isAudioJackPluggedIn(): Promise<boolean>

  addListener('AUDIO_CHANGED_NOTIFICATION', callback)
}
```

## Related Issues/Bugs
robinpowered/robin-rooms-react#801

### CODE REVIEW

File: android/src/main/java/com/robinpowered/RNAudioJack/RNAudioJackModule.java [hunk 1]
```diff
@@ -2,20 +2,83 @@
...
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.os.Build;
+import android.media.AudioDeviceInfo;
+import android.media.AudioManager;
 
-  private final ReactApplicationContext reactContext;
+public class RNAudioJackModule extends ReactContextBaseJavaModule {
+  private static final String MODULE_NAME = "RNAudioJack";
```
Comment (christophermark) [inline android/src/main/java/com/robinpowered/RNAudioJack/RNAudioJackModule.java]: Let's remove this prefix: `RN`.

When we call `NativeModules.AudioJack`, it's already in the React Native context. No reason to be redundant in the naming.

File: android/src/main/java/com/robinpowered/RNAudioJack/RNAudioJackModule.java [hunk 2]
```diff
@@ -2,20 +2,83 @@
...
   }
 
   @Override
   public String getName() {
-    return "RNAudioJack";
+    return MODULE_NAME;
+  }
+
+  @ReactMethod
+  public void isAudioJackPluggedIn(final Promise promise) {
```
Comment (christophermark) [inline android/src/main/java/com/robinpowered/RNAudioJack/RNAudioJackModule.java]: This naming is a bit redundant with the module name. Might be better as `getIsPluggedIn` or `isPluggedIn`:

```js
// JS
var audioJackPluggedIn = NativeModules.AudioJack.isPluggedIn();
```
vs 
```js
// JS
var audioJackPluggedIn = NativeModules.AudioJack.isAudioJackPluggedIn();
```

**Review by christophermark — CHANGES_REQUESTED** [review]
Some small stuff. Looking good here!

File: android/src/main/java/com/robinpowered/RNAudioJack/RNAudioJackModule.java (hunk 1 above)
Comment (atticoos) [reply android/src/main/java/com/robinpowered/RNAudioJack/RNAudioJackModule.java]: Little backstory on why you generally see `RN` in the prefix here: the origins come from iOS.

In iOS, files (classes) are not namespaced (unless you use frameworks, which isn't common). So you'll usually see al of `RCT*` or `RN*` prefixes. 

Again in iOS, when you create the module, you call a macro `RCT_EXPORT_MODULE()`, which creates a native module with the same name as the class name. And since the class name will use some sort of package identifier in its name ("RN"), it becomes the module name, and then you'd end up doing the same on Android to keep it consistent with the name generated by iOS.

HOWEVER -- you can now name your module in iOS with `RCT_EXPORT_MODULE("NAME")`.

----

Given that this is an open source project, we want to make sure the files are namespaced. So the iOS filename/class should continue to be `RNAudioJack`, but we can use `RCT_EXPORT_MODULE("AudioJack")` to create a friendly name for the module.

TLDR - remove it, but it was put there with good intentions

File: android/src/main/java/com/robinpowered/RNAudioJack/RNAudioJackModule.java [hunk 3]
```diff
@@ -2,20 +2,83 @@
...
+import android.content.IntentFilter;
+import android.os.Build;
+import android.media.AudioDeviceInfo;
+import android.media.AudioManager;
 
-  private final ReactApplicationContext reactContext;
+public class RNAudioJackModule extends ReactContextBaseJavaModule {
+  private static final String MODULE_NAME = "RNAudioJack";
+  private static final String AUDIO_CHANGED_NOTIFICATION = "AUDIO_CHANGED_NOTIFICATION";
+  private static final String IS_AUDIO_JACK_PLUGGED_IN = "isAudioJackPluggedIn";
```
Comment (atticoos) [inline android/src/main/java/com/robinpowered/RNAudioJack/RNAudioJackModule.java]: Echoing @christophermark's feedback in https://github.com/robinpowered/react-native-audio-jack/pull/2/files#r159991719 to do the same here, `isPluggedIn`

File: ios/RNAudioJack.m [hunk 4]
```diff
@@ -1,11 +1,76 @@
...
+    return NO;
+}
+
 - (dispatch_queue_t)methodQueue
 {
     return dispatch_get_main_queue();
 }
-RCT_EXPORT_MODULE()
+
+RCT_EXPORT_MODULE(@"AudioJack")
```
Comment (atticoos) [inline ios/RNAudioJack.m]: Minor nit, but prefer to put this right below `@implementation` at the top.

Otherwise it's hard to discover the name of the module, and helps show the fact that it's a module to begin with

File: android/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -1,4 +1,4 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-          package="com.reactlibrary">
+          package="com.robinpowered.RNAudioJack">
```
Comment (atticoos) [inline android/src/main/AndroidManifest.xml]: RE naming feedback, this, and other references in Android, should drop the `RN` prefix

File: android/src/main/java/com/robinpowered/AudioJack/AudioJackModule.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        data.putBoolean(IS_PLUGGED_IN, pluggedInState == 1);
+
+        if (reactContext.hasActiveCatalystInstance()) {
+          reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
+              .emit(AUDIO_CHANGED_NOTIFICATION, data);
+        }
+      }
+    };
+
+    reactContext.registerReceiver(headsetReceiver, headsetFilter);
```
Comment (atticoos) [inline android/src/main/java/com/robinpowered/AudioJack/AudioJackModule.java]: @christophermark do we need to unregister these during things like `onPause`?

**Review by christophermark — APPROVED** [review]
👍 Good work here!

**Review by atticoos — APPROVED** [review]
Well done 👍
